src/realm/sync/noinst/migration_store.cpp

```diff
@@ -68,61 +62,84 @@ MigrationStore::MigrationStore(DBRef db,
+        m_state = MigrationState::NotStarted;
+        tr->commit();
+    }
+    else {
+        auto migration_store_obj = migration_table->get_object(0);
+        m_state = static_cast<MigrationState>(migration_store_obj.get<int64_t>(m_migration_state));
+        m_query_string = migration_store_obj.get<String>(m_migration_query_str);
+    }
 }
 
 std::shared_ptr<realm::SyncConfig> MigrationStore::convert_sync_config(std::shared_ptr<realm::SyncConfig> config)
 {
     REALM_ASSERT(config);
 
     std::lock_guard lock{m_mutex};
-    if (config->flx_sync_requested) {
-        cancel_migration();
-        return config;
+    if (m_state == MigrationState::Completed && config->flx_sync_requested) {
+        // Migration state is no longer needed - natively using flexible sync
+        reset_migration_info();
+        return config; // return original sync config
     }
     if (m_state == MigrationState::NotStarted) {
+        // Using PBS and not migrated, return original sync config
         return config;
     }
 
+    // Using PBS and migrated, update copy of original sync config to FLX and return
     auto flx_config = std::make_shared<realm::SyncConfig>(*config); // deep copy
-    flx_config->partition_value = "";
+    flx_config->partition_value = {};
     flx_config->flx_sync_requested = true;
 
     return flx_config;
 }
 
-void MigrationStore::migrate_to_flx(std::string rql_query_string)
+void MigrationStore::migrate_to_flx(std::string_view rql_query_string)
 {
     REALM_ASSERT(!rql_query_string.empty());
 
-    std::unique_lock lock{m_mutex};
-    m_query_string = rql_query_string;
-    m_state = MigrationState::Completed;
-    lock.unlock();
-
-    auto tr = m_db->start_write();
-    auto migration_store_obj = tr->get_table(m_migration_table)->get_object(0);
-    migration_store_obj.set(m_migration_query_str, m_query_string);
-    migration_store_obj.set(m_migration_state, int64_t(m_state));
-    migration_store_obj.set(m_migration_completed_at, Timestamp{std::chrono::system_clock::now()});
-    tr->commit();
+    {
+        std::unique_lock lock{m_mutex};
+        m_query_string = rql_query_string;
+        m_state = MigrationState::Completed;
+
+        auto tr = m_db->start_write();
+        auto migration_store_obj = tr->get_table(m_migration_table)->get_object(0);
+        migration_store_obj.set(m_migration_query_str, m_query_string);
+        migration_store_obj.set(m_migration_state, int64_t(m_state));
+        migration_store_obj.set(m_migration_completed_at, Timestamp{std::chrono::system_clock::now()});
+        tr->commit();
+    }
 
-    m_on_migration_state_changed(m_state);
+    m_on_migration_state_changed(MigrationState::Completed);
 }
 
 void MigrationStore::cancel_migration()
 {
+    // Clear the migration state
+    reset_migration_info();
+
+    m_on_migration_state_changed(MigrationState::NotStarted);
+}
+
+void MigrationStore::reset_migration_info()
+{
+    std::lock_guard lock{m_mutex};
```

Comment:
I notice we both lock a mutex and hold a global write lock in all these mutator functions, is it just so that make_subscription() doesn't need to actually read anything from the db to be able to make subscriptions?

Reply:
yeah, we cache the values